graphql/provider.go

```diff
@@ -40,10 +68,67 @@ func graphqlConfigure(ctx context.Context, d *schema.ResourceData) (interface{},
 		GQLServerUrl:   d.Get("url").(string),
 		RequestHeaders: d.Get("headers").(map[string]interface{}),
 	}
+
+	if d.Get("login_query") != "" {
+		queryResponse, resBytes, err := queryExecute(ctx, d, config, "login_query", "login_query_variables")
+		if err != nil {
+			return nil, diag.FromErr(fmt.Errorf("unable to execute read query: %w", err))
+		}
+
+		if queryErrors := queryResponse.ProcessErrors(); queryErrors.HasError() {
+			return nil, *queryErrors
+		}
+
+		var queryResponseCty cty.Value
+		if queryResponseCty, err = gocty.ToCtyValue(string(resBytes), cty.String); err != nil {
+			return nil, diag.FromErr(err)
+		}
+
+		evalCtx := &hcl.EvalContext{
+			Variables: map[string]cty.Value{
+				"login_query_response": queryResponseCty,
+			},
+			Functions: map[string]function.Function{
+				"jsondecode": stdlib.JSONDecodeFunc,
+				//TODO: maybe add other useful functions
+			},
+		}
+
+		config.RequestAuthenticatedHeaders = make(map[string]interface{})
+		for k, v := range d.Get("authenticated_headers").(map[string]interface{}) {
+			var expr hcl.Expression
+			var hclDiags hcl.Diagnostics
+			if expr, hclDiags = hclsyntax.ParseTemplate([]byte(v.(string)), k, hcl.InitialPos); len(hclDiags) > 0 {
+				return nil, convertDiagnosticsFromHCLToTerraformSDK(hclDiags)
+			}
+
+			var interpolated cty.Value
+			if interpolated, hclDiags = expr.Value(evalCtx); len(hclDiags) > 0 {
+				return nil, convertDiagnosticsFromHCLToTerraformSDK(hclDiags)
+			}
+			config.RequestAuthenticatedHeaders[k] = interpolated.AsString()
+		}
+	}
+
 	return config, diag.Diagnostics{}
 }
 
 type graphqlProviderConfig struct {
 	GQLServerUrl   string
 	RequestHeaders map[string]interface{}
+
+	RequestAuthenticatedHeaders map[string]interface{}
+}
+
+func convertDiagnosticsFromHCLToTerraformSDK(hclDiags hcl.Diagnostics) diag.Diagnostics {
+	diags := make(diag.Diagnostics, len(hclDiags))
+	for i, hclDiag := range hclDiags {
+		// HCL severity enum is: 0 (invalid), 1 (error), 2 (warning)
+		// terraform SDK severity enum is: 0 (error), 1 (warning)
+		diags[i].Severity = diag.Severity(hclDiag.Severity - 1)
+		diags[i].Summary = hclDiag.Summary
+		diags[i].Detail = hclDiag.Detail
+		// diags[i].AttributePath = ?
```

Comment:
This may be helpful to end users when troubleshooting invalid expressions.

Reply:
One of these days I will find time to refactor this over to the new SDK.

Reply:
I do not think it would resolve this specific issue as [HCL v2](https://github.com/hashicorp/hcl/blob/v2.11.1/diagnostic.go#L26) and [Terraform SDK v2](https://github.com/hashicorp/terraform-plugin-sdk/blob/v2.10.1/diag/diagnostic.go#L44) still differ in their diagnostic types:
- https://github.com/hashicorp/hcl/blob/v2.11.1/diagnostic.go#L26
- https://github.com/hashicorp/terraform-plugin-sdk/blob/v2.10.1/diag/diagnostic.go#L44

Anyway, I can submit another PR to migrate to the new SDK as I've already done this for other providers.

Reply:
Ah, indeed, got it :sweat_smile: 

Same issue about diagnostic types though, still differs from HCL: https://github.com/hashicorp/terraform-plugin-framework/blob/v0.5.0/diag/diagnostic.go#L18
However, there is no `AttributePath` in that version so it'll solve that conversation altogether :)